.buildkite/e2e/pipeline-gen/main.go

```diff
@@ -0,0 +1,390 @@
+	EnvVarPipeline             = "PIPELINE"
+	EnvVarClusterName          = "CLUSTER_NAME"
+	EnvVarTestOpts             = "TEST_OPTS"
+	EnvVarTestsMatch           = "TESTS_MATCH"
+	EnvVarBuildLicensePubkey   = "BUILD_LICENSE_PUBKEY"
+	EnvVarLicensePubKey        = "export LICENSE_PUBKEY"
+	EnvVarTestLicense          = "TEST_LICENSE"
+	EnvVarMonitoringSecrets    = "MONITORING_SECRETS"
+	EnvVarE2EJson              = "E2E_JSON"
+	EnvVarGoTags               = "GO_TAGS"
+	EnvVarOperatorImage        = "OPERATOR_IMAGE"
+	EnvVarE2EImage             = "E2E_IMG"
+)
+
+var (
+	//go:embed pipeline.tpl.yaml
+	pipelineTemplate string
+
+	// providersInDocker are k8s providers that require the deployer to run in Docker
+	providersInDocker = []string{"kind", "aks", "ocp", "tanzu"}
+	// providersNoCleanup are k8s providers that do not require the cluster to be deleted after use
+	providersNoCleanup = []string{"kind"}
+
+	semverRE = regexp.MustCompile(`\d*\.\d*\.\d*(-\w*)?`)
+	chars    = []rune("0123456789abcdefghijklmnopqrstuvwxyz")
+
+	shortcuts = map[string]string{
+		"p": EnvVarProvider,
+		"k": EnvVarK8sVersion,
+		"s": EnvVarStackVersion,
+		"t": EnvVarTestsMatch,
+	}
+
+	fixed string
+	mixed string
+
+	output string
+
+	rootDir string
+)
+
+func init() {
+	flag.StringVar(&fixed, "f", "", "Fixed variables (comma-separated list)")
+	flag.StringVar(&mixed, "m", "", "Mixed variables (comma-separated list of colon-separated list)")
+	flag.StringVar(&output, "o", "buildkite-pipeline", "Type of output: buildkite-pipeline or envfile")
+	flag.Parse()
+
+	rand.Seed(time.Now().UTC().UnixNano())
+
+	_, filename, _, ok := runtime.Caller(0)
+	if !ok {
+		fmt.Printf("failed to get current path")
+		os.Exit(1)
+	}
+	rootDir = filepath.Join(filepath.Dir(filename), "../../..")
+}
+
+func main() {
+	stat, err := os.Stdin.Stat()
+	handlErr("Failed to read stdin", err)
+
+	var groups []Group
+
+	// no stdin
+	if stat.Mode()&os.ModeCharDevice != 0 {
+
+		fixedEnv, err := stringListToEnv(fixed)
+		handlErr("Failed to read fixed variables", err)
+
+		mixedEnv, err := stringListToEnvs(mixed)
+		handlErr("Failed to read mixed variables", err)
+
+		groups = []Group{{
+			Fixed: fixedEnv,
+			Mixed: mixedEnv,
+		}}
+
+	} else {
+		in, err := io.ReadAll(os.Stdin)
+		handlErr("Failed to read stdin", err)
+		if len(in) == 0 {
+			handlErr("Failed to read stdin", errors.New("nothing on /dev/stdin"))
+		}
+
+		err = yaml.Unmarshal(in, &groups)
+		handlErr("Failed to parse stdin", err)
+	}
+
+	// build a flat list of the tests to run
+	tests := make([]TestsSuiteRun, 0)
+	cleanup := false
+	for i := range groups {
+		if groups[i].Mixed == nil {
+			groups[i].Mixed = []Env{{}}
+		}
+		for j := range groups[i].Mixed {
+			ts, err := newTestsSuite(groups[i].Label, groups[i].Fixed, len(groups[i].Mixed), groups[i].Mixed[j])
+			handlErr("Failed to create tests suite", err)
+
+			tests = append(tests, ts)
+			cleanup = cleanup || ts.Cleanup
+		}
+	}
+
+	if output == "envfile" {
```

Comment:
Should we validate that the chosen `output` is a supported value. If I put anything but `envfile` I get buildkite which is fine but then maybe it should just be a `-e|--envfile` flag?

Reply:
I like the `-e|--envfile` that avoids the validation.